File: stock-advisor/src/advisor/verdict.py
"""The decision engine: evidence in, one action out, through a fixed sequence.

Nine stages, always in this order. The order is the product — it is what makes
the answer reproducible and auditable instead of a judgement call:

    0 Evidence      what is actually known; caps confidence
    1 Vetoes        fraud, default, ruinous leverage — override everything
    2 Position      a breached exit price is a decision already made
    3 Business      is it still the company you thought you owned?
    4 Valuation     what are you paying for that business?
    5 News          has anything material changed?
    6 Trend         what is price doing?
    7 Regime        does this market reward new risk?
    8 Horizon       long-term (business) or short-term (trend)?
    9 Synthesis     action, conviction, levels, and the chain that produced them

Conflicts are resolved differently by book, because the two books are playing
different games (owner's instruction):

    INVESTING  fundamentals decide, the trend only times the entry. A cheap,
               sound business in a downtrend is a "wait", not an "avoid".
    TRADING    a genuine conflict returns NO POSITION with both cases stated.
               In a book that lives by the exit price, ambiguity is a reason to
               stand aside, not to pick a side.

Nothing here estimates a missing figure. An unknown stays unknown and lowers
the confidence ceiling instead.
"""
from __future__ import annotations

from dataclasses import dataclass, field, asdict

ACTIONS = ("SELL", "TRIM", "HOLD", "WATCH", "AVOID", "ACCUMULATE", "BUY")


@dataclass
class Step:
    stage: str
    finding: str


@dataclass
class Verdict:
    symbol: str
    action: str
    conviction: int
    horizon: str
    chain: list[Step] = field(default_factory=list)
    levels: dict | None = None
    short_case: bool = False
    book: str = "investing"
    conflict: str | None = None
    unknowns: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        d = asdict(self)
        d["chain"] = [asdict(s) if not isinstance(s, dict) else s for s in self.chain]
        return d


def _levels(features: dict | None) -> dict | None:
    if not features or not features.get("atr14"):
        return None
    last = features["close"]
    stop = max(features.get("swing_low_20d", 0.0), last - 2 * features["atr14"])
    if stop >= last:
        stop = last - 2 * features["atr14"]
    risk = last - stop
    if risk <= 0:
        return None
    return {"reference_price": round(last, 2), "stop": round(stop, 2),
            "target1": round(last + 1.5 * risk, 2), "target2": round(last + 2.5 * risk, 2),
            "add_below": round(last * 0.94, 2)}


def decide(symbol: str, features: dict | None = None, fundamentals: dict | None = None,
           fundamental_score: float | None = None, news_items: list | None = None,
           news_pressure: dict | None = None, stage: int | None = None,
           regime_risk_on: bool = True, holding: dict | None = None,
           book: str = "investing", valuation: dict | None = None) -> Verdict:
    """Run the nine stages. Every argument may be None — missing evidence lowers
    the confidence ceiling rather than being filled in."""
    chain: list[Step] = []
    unknowns: list[str] = []
    conviction = 50.0

    def step(stage_name: str, finding: str) -> None:
        chain.append(Step(stage_name, finding))

    # --- 0: evidence -------------------------------------------------------
    have_price = features is not None
    have_fund = fundamentals is not None and fundamental_score is not None
    have_news = news_pressure is not None
    if not have_fund:
        unknowns.append("fundamentals")
    if not have_news:
        unknowns.append("news")
    if not have_price:
        unknowns.append("price_history")
    ceiling = 60 + (20 if have_fund else 0) + (20 if have_news else 0)
    step("Evidence", "Working from " + ", ".join(filter(None, [
        "price and trend" if have_price else "no price history",
        "financials" if have_fund else "no financials",
        "recent news" if have_news else "no news",
    ])) + f". Confidence is capped at {ceiling} because of what is missing.")

    def finish(action: str, conv: float, horizon: str, short_case: bool = False,
               conflict: str | None = None) -> Verdict:
        return Verdict(symbol=symbol, action=action,
                       conviction=int(max(5, min(ceiling, round(conv)))),
                       horizon=horizon, chain=chain, levels=_levels(features),
                       short_case=short_case, book=book, conflict=conflict,
                       unknowns=unknowns)

    # --- 1: hard vetoes ----------------------------------------------------
    veto = None
    if news_items:
        grave = [i for i in news_items
                 if getattr(i, "weight", i.get("weight", 0) if isinstance(i, dict) else 0) <= -3]
        if grave:
            g = grave[0]
            label = getattr(g, "event_label", None) or (g.get("event_label") if isinstance(g, dict) else "")
            title = getattr(g, "title", None) or (g.get("title") if isinstance(g, dict) else "")
            veto = f"A grave event is reported: {label.lower()} — “{title[:110]}”"
    if veto is None and fundamentals:
        de = fundamentals.get("debt_to_equity")
        cover = fundamentals.get("interest_cover")
        if de is not None and cover is not None and de > 3 and cover < 1.5:
            veto = (f"Debt is {de:.2f}× equity and operating profit covers interest only "
                    f"{cover:.1f}× — the balance sheet, not the business, now decides the outcome.")
    if veto:
        step("Veto", veto)
        return finish("SELL" if holding else "AVOID", 85,
                      "Long-term risk — this is about survival, not price")
    step("Veto check", "No fraud, default or ruinous-leverage flag found in the evidence held.")

    # --- 2: position state -------------------------------------------------
    forced_action = None
    if holding:
        stop = holding.get("stop")
        if stop and features and features["close"] <= stop:
            step("Position", f"Price {features['close']:.2f} is at or below the exit price "
                             f"of {stop:.2f} set when you bought.")
            return finish("SELL", 90,
                          "The decision was made when you were calm; this is the execution")
        weight = holding.get("weight")
        if weight is not None and weight > 0.15:
            step("Position", f"This position is {weight:.0%} of the portfolio — past the "
                             "15% concentration limit.")
            forced_action = "TRIM"

    # --- 3: business quality ----------------------------------------------
    quality = fundamental_score if have_fund else None
    if quality is not None:
        bits = [f"Fundamental score {quality:.0f}/100"]
        if fundamentals.get("roce") is not None:
            bits.append(f"ROCE {fundamentals['roce']:.1%}")
        elif fundamentals.get("roe") is not None:
            bits.append(f"ROE {fundamentals['roe']:.1%}")
        if fundamentals.get("debt_to_equity") is not None:
            bits.append(f"debt/equity {fundamentals['debt_to_equity']:.2f}")
        step("Business", ", ".join(bits) + ".")
        conviction += 12 if quality >= 70 else 4 if quality >= 55 else -6 if quality >= 40 else -18
    else:
        step("Business", "No financials available, so nothing here can be a long-term call.")

    # --- 4: valuation ------------------------------------------------------
    # A discounted-cash-flow view outranks a bare multiple when it exists: the
    # multiple tells you what you pay, the DCF tells you what you get.
    val_in = valuation if isinstance(valuation, dict) else None   # capture before rebinding
    valuation = "unknown"                                          # now the category
    if val_in and val_in.get("margin_of_safety") is not None:
        mos = val_in["margin_of_safety"]
        valuation = ("cheap" if mos > 0.15 else "fair" if mos > -0.15
                     else "rich" if mos > -0.35 else "extreme")
        bits = [f"Fair value about {val_in.get('fair_value_base')}, "
                f"a margin of safety of {mos:+.0%} — {valuation}"]
        if val_in.get("implied_growth") is not None:
            bits.append(f"the price already assumes {val_in['implied_growth']:.1%} "
                        "annual cash-flow growth")
        step("Valuation", "; ".join(bits) + ".")
        conviction += {"cheap": 12, "fair": 3, "rich": -8, "extreme": -16}[valuation]
    elif fundamentals and fundamentals.get("pe") and fundamentals["pe"] > 0:
        pe = fundamentals["pe"]
        valuation = ("cheap" if pe < 18 else "fair" if pe < 35
                     else "rich" if pe < 60 else "extreme")
        step("Valuation", f"P/E of {pe:.1f} — {valuation}. (No cash-flow valuation "
                          "available, so this is what you pay, not what you get.)")
        conviction += {"cheap": 10, "fair": 3, "rich": -6, "extreme": -14}[valuation]
    else:
        unknowns.append("valuation")

    # --- 5: news overlay ---------------------------------------------------
    tone = "none"
    if have_news:
        tone = news_pressure.get("tone", "mixed")
        material = news_pressure.get("material_count", 0)
        step("News", (f"{material} material item(s) recently; net tone {tone}."
                      if material else "Nothing material in the recent window."))
        net = news_pressure.get("net", 0)
        conviction += 8 if net > 1.5 else -12 if net < -1.5 else 0

    # --- 6: trend ----------------------------------------------------------
    st = stage if stage is not None else 1
    trend_word = {2: "uptrend", 1: "basing", 3: "topping", 4: "downtrend"}.get(st, "unclear")
    detail = f"Price is in a {trend_word}"
    if features and features.get("rsi14") is not None:
        detail += f", RSI {features['rsi14']:.0f}"
    if features and features.get("dist_52w_high") is not None:
        detail += f", {features['dist_52w_high']:+.1%} from its 52-week high"
    step("Trend", detail + ".")
    conviction += {2: 10, 1: 0, 3: -8, 4: -16}.get(st, 0)
    rsi = (features or {}).get("rsi14")
    overbought = rsi is not None and rsi > 78
    oversold = rsi is not None and rsi < 25

    # --- 7: regime ---------------------------------------------------------
    if not regime_risk_on:
        step("Market", "The market is not in a clean uptrend, so new buying is restrained "
                       "regardless of how good the stock looks.")
        conviction -= 8

    # --- 8: horizon --------------------------------------------------------
    if quality is not None and quality >= 60 and valuation != "extreme":
        horizon = "Long term — the case rests on the business compounding; measure it in years"
    elif st == 2 and (quality is None or quality >= 40):
        horizon = "Short term — the case rests on the trend; it lives and dies by the exit price"
    else:
        horizon = "No horizon qualifies — neither the business nor the trend supports a position"
    step("Horizon", horizon)

    # --- 8b: conflict detection --------------------------------------------
    # A conflict is the models disagreeing about direction, not merely being
    # lukewarm: cheap-and-falling, or expensive-and-rising.
    conflict = None
    cheap = valuation in ("cheap",)
    dear = valuation in ("rich", "extreme")
    sound = quality is not None and quality >= 55
    weak_trend = st in (3, 4)
    strong_trend = st == 2
    if cheap and sound and weak_trend:
        conflict = ("The numbers say cheap and sound; the price says the market "
                    "disagrees. One of them is wrong, and the market usually knows "
                    "something first.")
    elif dear and strong_trend:
        conflict = ("The trend is strong but the price already assumes a great deal. "
                    "You would be buying momentum, not value.")

    if conflict:
        step("Conflict", conflict)
        if book == "trading":
            # In the trading book, ambiguity is a reason to stand aside: this book
            # lives by the exit price, and a contested thesis has no clean one.
            step("Resolution", "Trading book: a genuine conflict returns no position. "
                               "Both cases are stated above; neither is adopted.")
            return finish("HOLD" if holding else "WATCH", min(conviction, 40),
                          "No horizon — the models disagree and this book does not "
                          "take contested trades", conflict=conflict)
        step("Resolution", "Investing book: the business decides and the trend only "
                           "times the entry. A sound, cheap business in a downtrend "
                           "is a wait, not an avoid.")

    # --- 9: synthesis ------------------------------------------------------
    action = forced_action
    if action is None:
        bad = quality is not None and quality < 45
        good = quality is not None and quality >= 65
        if st == 4 and (bad or tone == "negative"):
            action = "SELL" if holding else "AVOID"
        elif st == 4 and book == "investing" and good and valuation in ("cheap", "fair"):
            # fundamentals decide, the trend only times: wait for the turn
            action = "HOLD" if holding else "WATCH"
        elif st == 4:
            action = "HOLD" if holding else "WATCH"
        elif holding:
            if tone == "negative" and bad:
                action = "SELL"
            elif overbought and valuation == "extreme":
                action = "TRIM"
            else:
                action = "HOLD"
        else:
            if good and st == 2 and regime_risk_on and not overbought:
                action = "BUY"
            elif good and (valuation == "cheap" or oversold):
                action = "ACCUMULATE"
            elif st == 2 and regime_risk_on and quality is None:
                action = "BUY"
            elif bad:
                action = "AVOID"
            else:
                action = "WATCH"

    short_case = st == 4 and (tone == "negative" or (quality is not None and quality < 40))
    if short_case and not holding:
        step("Short?", "The bear case is real, but a naked short has unlimited loss and Indian "
                       "cash-market shorts must be squared off the same day. Use a defined-risk "
                       "bear put spread sized to 0.5–1% of capital — never a naked short.")
    return finish(action, conviction, horizon, short_case, conflict)

File: live-console/public/verdict.js
/* The decision engine: turns evidence into one action, through a fixed sequence
   of gates. The sequence matters more than any single rule — it is what makes
   the answer reproducible instead of a vibe.

   Stage 0  Evidence inventory      what do we actually know? sets the confidence ceiling
   Stage 1  Hard vetoes             fraud, default, ruinous leverage — override everything
   Stage 2  Position state          a breached exit price is already a decision you made
   Stage 3  Business quality        is it still the company you thought you owned?
   Stage 4  Valuation               what are you paying for that business?
   Stage 5  News overlay            has anything material changed in the last two weeks?
   Stage 6  Trend and timing        what is price actually doing?
   Stage 7  Market regime           is this a market that rewards new risk?
   Stage 8  Horizon                 is the case long-term (business) or short-term (momentum)?
   Stage 9  Synthesis               action, conviction, levels, and the reasoning chain
*/
(function () {
  "use strict";

  function el(id) { return document.getElementById(id); }
  function esc(s) { return String(s == null ? "" : s).replace(/[&<>"]/g, function (c) {
    return { "&":"&amp;", "<":"&lt;", ">":"&gt;", '"':"&quot;" }[c]; }); }
  function num(n, d) { return n == null || isNaN(n) ? "—" :
    Number(n).toLocaleString("en-IN", { maximumFractionDigits: d == null ? 2 : d,
                                        minimumFractionDigits: d == null ? 2 : d }); }
  function pct(x, d) { return x == null || isNaN(x) ? "—" :
    (x >= 0 ? "+" : "") + (x * 100).toFixed(d == null ? 1 : d) + "%"; }

  var ACTIONS = {
    SELL:        { label: "Sell", cls: "down", order: 0 },
    TRIM:        { label: "Trim", cls: "warn", order: 1 },
    HOLD:        { label: "Hold", cls: "mut", order: 2 },
    WATCH:       { label: "Watch, don't buy yet", cls: "mut", order: 3 },
    AVOID:       { label: "Avoid", cls: "down", order: 4 },
    ACCUMULATE:  { label: "Buy on dips", cls: "sky", order: 5 },
    BUY:         { label: "Buy", cls: "up", order: 6 },
    HEDGE:       { label: "Hedge, don't short", cls: "warn", order: 7 }
  };

  /* ---------- the engine ---------- */
  function decide(input) {
    var q = input.quote, f = input.fundamentals, news = input.news,
        held = input.held, regimeRiskOn = input.regimeRiskOn;
    var chain = [];           // the reasoning, in the order it was applied
    var conviction = 50;      // 0-100, adjusted by each stage
    var action = null, horizon = null;

    function step(stage, finding) { chain.push({ stage: stage, finding: finding }); }

    /* -- Stage 0: evidence inventory -- */
    var have = { price: !!q, fundamentals: !!f && !f.error, news: !!news && !news.error };
    var ceiling = 60 + (have.fundamentals ? 20 : 0) + (have.news ? 20 : 0);
    step("Evidence", "Working from " +
      [have.price ? "price and trend" : null,
       have.fundamentals ? "financials" : "no financials",
       have.news ? "recent news" : "no news feed"].filter(Boolean).join(", ") +
      ". Confidence is capped at " + ceiling + " because of what is missing.");

    /* -- Stage 1: hard vetoes -- */
    var veto = null;
    if (have.news) {
      var grave = (news.items || []).filter(function (i) {
        return i.weight <= -3 && (i.ageDays == null || i.ageDays <= 21); });
      if (grave.length) veto = "A grave event is in the news: " + grave[0].eventLabel.toLowerCase() +
        " — “" + grave[0].title.slice(0, 110) + "”";
    }
    if (!veto && have.fundamentals && f.debtToEquity != null && f.debtToEquity > 3 &&
        f.interestCover != null && f.interestCover < 1.5) {
      veto = "Debt is " + num(f.debtToEquity) + "× equity and operating profit covers interest only " +
        num(f.interestCover, 1) + "× — the balance sheet, not the business, now decides the outcome.";
    }
    if (veto) {
      step("Veto", veto);
      return finish(held ? "SELL" : "AVOID", 85,
        "long-term risk — this is about survival, not price", chain, input);
    }
    step("Veto check", "No fraud, default or ruinous-leverage flag.");

    /* -- Stage 2: position state (holdings only) -- */
    if (held) {
      if (held.stop && q && q.last <= held.stop) {
        step("Position", "Price " + num(q.last) + " is at or below the exit price of " +
          num(held.stop) + " you set when you bought.");
        return finish("SELL", 90,
          "the decision was made when you were calm; this is only the execution", chain, input);
      }
      if (held.weight != null && held.weight > 0.15) {
        step("Position", "This position is " + pct(held.weight, 0) +
          " of the portfolio — past the 15% concentration limit.");
        action = "TRIM";
      }
    }

    /* -- Stage 3: business quality -- */
    var quality = null;
    if (have.fundamentals && input.fundScore != null) {
      quality = input.fundScore;
      step("Business", "Fundamental score " + quality + "/100" +
        (f.roce != null ? ", ROCE " + pct(f.roce) : f.roe != null ? ", ROE " + pct(f.roe) : "") +
        (f.debtToEquity != null ? ", debt/equity " + num(f.debtToEquity) : "") + ".");
      conviction += quality >= 70 ? 12 : quality >= 55 ? 4 : quality >= 40 ? -6 : -18;
    } else {
      step("Business", "No financials available, so nothing here can be a long-term call.");
    }

    /* -- Stage 4: valuation -- */
    var valuation = "unknown";
    if (have.fundamentals && f.pe != null && f.pe > 0) {
      valuation = f.pe < 18 ? "cheap" : f.pe < 35 ? "fair" : f.pe < 60 ? "rich" : "extreme";
      step("Valuation", "P/E of " + num(f.pe, 1) + " — " + valuation + ".");
      conviction += valuation === "cheap" ? 10 : valuation === "fair" ? 3
        : valuation === "rich" ? -6 : -14;
    }

    /* -- Stage 5: news overlay -- */
    var newsTone = "none";
    if (have.news && news.pressure) {
      newsTone = news.pressure.tone;
      var material = (news.items || []).filter(function (i) { return Math.abs(i.weight) >= 2; });
      step("News", material.length
        ? material.length + " material item(s) in the last two weeks; net tone " + newsTone +
          ". Most significant: “" + material[0].title.slice(0, 100) + "”"
        : "Nothing material in the last two weeks.");
      conviction += news.pressure.net > 1.5 ? 8 : news.pressure.net < -1.5 ? -12 : 0;
    }

    /* -- Stage 6: trend and timing -- */
    var st = input.stage, trendWord = { 2:"uptrend", 1:"basing", 3:"topping", 4:"downtrend" }[st];
    step("Trend", "Price is in a " + trendWord +
      (q && q.rsi14 != null ? ", RSI " + num(q.rsi14, 0) : "") +
      (q && q.high52 ? ", " + pct(q.last / q.high52 - 1) + " from its 52-week high" : "") + ".");
    conviction += st === 2 ? 10 : st === 1 ? 0 : st === 3 ? -8 : -16;
    var overbought = q && q.rsi14 != null && q.rsi14 > 78;
    var oversold = q && q.rsi14 != null && q.rsi14 < 25;

    /* -- Stage 7: regime -- */
    if (!regimeRiskOn) {
      step("Market", "The market is not in a clean uptrend, so new buying is restrained " +
        "regardless of how good the stock looks.");
      conviction -= 8;
    }

    /* -- Stage 8: horizon -- */
    if (quality != null && quality >= 60 && valuation !== "extreme") {
      horizon = "Long term — the case rests on the business compounding, so measure it in years";
    } else if (st === 2 && (quality == null || quality >= 40)) {
      horizon = "Short term — the case rests on the trend, so it lives and dies by the exit price";
    } else {
      horizon = "No horizon qualifies — neither the business nor the trend supports a position";
    }
    step("Horizon", horizon);

    /* -- Stage 9: synthesis -- */
    if (!action) {
      var badBusiness = quality != null && quality < 45;
      var goodBusiness = quality != null && quality >= 65;
      if (st === 4 && (badBusiness || newsTone === "negative")) {
        action = held ? "SELL" : "AVOID";
      } else if (st === 4) {
        action = held ? "HOLD" : "WATCH";
      } else if (held) {
        if (newsTone === "negative" && badBusiness) action = "SELL";
        else if (overbought && valuation === "extreme") action = "TRIM";
        else action = "HOLD";
      } else {
        if (goodBusiness && st === 2 && regimeRiskOn && !overbought) action = "BUY";
        else if (goodBusiness && (valuation === "cheap" || oversold)) action = "ACCUMULATE";
        else if (st === 2 && regimeRiskOn && quality == null) action = "BUY";
        else if (badBusiness) action = "AVOID";
        else action = "WATCH";
      }
    }

    // The "short it" question, answered honestly rather than ignored.
    var shortCase = (st === 4 && (newsTone === "negative" || (quality != null && quality < 40)));
    if (shortCase && !held) {
      step("Short?", "The bear case is real, but a naked short has unlimited loss and, in India, " +
        "cash-market shorts must be squared off the same day. If you want this exposure, use a " +
        "defined-risk bear put spread sized to 0.5–1% of capital — never a naked short.");
    }

    return finish(action, Math.round(Math.max(5, Math.min(ceiling, conviction))),
      horizon, chain, input, shortCase);
  }

  function finish(action, conviction, horizon, chain, input, shortCase) {
    var q = input.quote, levels = null;
    if (q && q.atr14) {
      var stop = Math.max(q.swingLow20 || 0, q.last - 2 * q.atr14);
      if (stop >= q.last) stop = q.last - 2 * q.atr14;
      var risk = q.last - stop;
      levels = { entry: q.last, stop: stop, t1: q.last + 1.5 * risk, t2: q.last + 2.5 * risk,
                 addBelow: q.last * 0.94 };
    }
    return { action: action, conviction: conviction, horizon: horizon,
             chain: chain, levels: levels, shortCase: !!shortCase, symbol: input.symbol };
  }

  window.ADV_VERDICT = { decide: decide, ACTIONS: ACTIONS };

  /* ---------- rendering ---------- */
  function card(v, extra) {
    var a = ACTIONS[v.action] || ACTIONS.WATCH;
    var lv = v.levels;
    return '<div class="card" style="margin-bottom:14px">' +
      '<div style="display:flex;align-items:baseline;gap:12px;flex-wrap:wrap;margin-bottom:8px">' +
      '<span class="sym" style="font-size:17px">' + esc(v.symbol) + '</span>' +
      '<span class="pill ' + a.cls + '" style="font-size:12px;padding:5px 12px">' + a.label + '</span>' +
      '<span class="muted">conviction ' + v.conviction + '/100</span>' +
      (extra ? '<span class="muted">' + extra + '</span>' : "") + '</div>' +
      '<p style="margin:0 0 10px"><strong>' + esc(v.horizon) + '</strong></p>' +
      '<details><summary style="cursor:pointer;font-family:Archivo,sans-serif;font-weight:700;' +
      'font-size:12.5px;color:var(--muted)">How this was decided — ' + v.chain.length +
      ' steps</summary><ol style="margin:10px 0 0;padding-left:20px;font-size:13.5px">' +
      v.chain.map(function (c) {
        return '<li style="margin-bottom:6px"><strong>' + esc(c.stage) + ':</strong> ' +
          esc(c.finding) + '</li>'; }).join("") + '</ol></details>' +
      (lv ? '<div class="rowline" style="margin-top:12px;gap:18px;font-size:13px">' +
        '<span class="muted">Exit price <strong class="mono" style="color:var(--down)">' +
        num(lv.stop) + '</strong></span>' +
        '<span class="muted">Target <strong class="mono" style="color:var(--up)">' +
        num(lv.t1) + '</strong> then <strong class="mono" style="color:var(--up)">' +
        num(lv.t2) + '</strong></span>' +
        (v.action === "ACCUMULATE" ? '<span class="muted">Add below <strong class="mono">' +
          num(lv.addBelow) + '</strong></span>' : "") + '</div>' : "") +
      (v.shortCase ? '<div class="notice warnn" style="margin-top:10px">The bear case is real. ' +
        'If you want to act on it, use a defined-risk bear put spread — never a naked short.</div>' : "") +
      '</div>';
  }
  window.ADV_VERDICT.card = card;
})();
